### ISSUE TITLE
Concurrent updates behave badly

### ISSUE THREAD

**Issue by sopoforic** [burtonator/polar-bookshelf#1492]
2.0 beta 25 on Chrome 84

Polar handles concurrent updates poorly, potentially causing data loss.

# Steps to reproduce
1. Open polar in two tabs, A and B
2. In tab A, select a document X and press the tag button
3. In tab B, press the archive button on X
4. In tab A, add a tag and press Done
5. Now the document is unarchived.

It appears that updating tags is actually replacing the whole state of the document. This can also cause **loss of annotations**:

# Steps to reproduce, redux
1. Open polar in tab A.
2. Double click X, opening it in a new tab, B.
3. Click the tag button on document X in tab A.
4. Make some highlights, etc. in tab B. Allow them to sync.
5. Add a tag in tab A and press Done.
6. See that the annotations have vanished in tab B.

It would be dangerous to use polar on multiple devices, if this kind of thing can happen.

**Comment by sopoforic** [comment]
Note that this happened even though the data had already synced fully between tabs. The tag box seems to be taking a copy of the doc state when it is opened, leading to this problem.

**Comment by burtonator** [comment]
Ok... so the concurrency issue is designed to auto-update all tabs so there will ALWAYS be a race.

BUT.. seems you have a legit but while the tag button is open in the document repository?  If it's not reading the most recent version of the doc that would explain this problem.  Let me know if that's the issue and I'll investigate.

Good find.

**Comment by burtonator** [comment]
and this was from the doc repo correct?

**Comment by sopoforic** [comment]
Yes, I opened the tag box from the doc repo. I didn't test whether I'd get similar results with multiple copies open in the PDF viewer.

**Comment by sopoforic** [comment]
And regarding "always be a race", yes, but if the updates affected only the bit of metadata being changed, you wouldn't see this kind of problem. A tag update could conflict with--and overwrite--tags changed in another session, but it shouldn't cause annotations to disappear.

Even if all updates are on the whole document's metadata, the right way to handle this would probably be to always fetch the latest version on clicking Done and then merge in the changes from the tag box.

**Comment by burtonator** [comment]
I'm trying to take an optimistic locking model to ensure that the user is using the most up to date copy of the data, otherwise the UI will be slower waiting for data all the time.

The latency is like 4-5 seconds so probably wouldn't happen in reality.

The longer term goal is to have everything like a CRDT where we do mutations on the server such that it 'merges' the result, not overwrites it.
